godot-macros/src/util/mod.rs

```diff
@@ -0,0 +1,205 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+// Note: some code duplication with codegen crate
+
+use crate::ParseResult;
+use proc_macro2::{Ident, TokenTree};
+use quote::format_ident;
+use quote::spanned::Spanned;
+use venial::{Error, Function, GenericParamList, Impl, WhereClause};
+
+mod kv_parser;
+
+pub(crate) use kv_parser::KvParser;
+
+pub fn ident(s: &str) -> Ident {
+    format_ident!("{}", s)
+}
+
+pub fn bail_fn<R, T>(msg: impl AsRef<str>, tokens: T) -> ParseResult<R>
+where
+    T: Spanned,
+{
+    Err(error_fn(msg, tokens))
+}
+
+macro_rules! bail {
+    ($tokens:expr, $format_string:literal $($rest:tt)*) => {
+        $crate::util::bail_fn(format!($format_string $($rest)*), $tokens)
+    }
+}
+
+pub(crate) use bail;
+
+pub fn error_fn<T>(msg: impl AsRef<str>, tokens: T) -> Error
+where
+    T: Spanned,
+{
+    Error::new_at_span(tokens.__span(), msg.as_ref())
+}
+
+macro_rules! error {
+    ($tokens:expr, $format_string:literal $($rest:tt)*) => {
+        $crate::util::error_fn(format!($format_string $($rest)*), $tokens)
+    }
+}
+
+pub(crate) use error;
+
+pub fn reduce_to_signature(function: &Function) -> Function {
+    let mut reduced = function.clone();
+    reduced.vis_marker = None; // TODO needed?
+    reduced.attributes.clear();
+    reduced.tk_semicolon = None;
+    reduced.body = None;
+
+    reduced
+}
+
+fn is_punct(tt: &TokenTree, c: char) -> bool {
+    match tt {
+        TokenTree::Punct(punct) => punct.as_char() == c,
+        _ => false,
+    }
+}
+
+// ----------------------------------------------------------------------------------------------------------------------------------------------
+// Validation for trait/impl
+
+/// Given an impl block for a trait, returns whether that is an impl
+/// for a trait with the given name.
+///
+/// That is, if `name` is `"MyTrait"`, then this function returns true
+/// if and only if `original_impl` is a declaration of the form `impl
+/// MyTrait for SomeType`. The type `SomeType` is irrelevant in this
+/// example.
```

Comment:
I'm not sure if this is new code since Git doesn't detect it as rename, but I'll mention it anyway:

Lines should be wider. The horizontal line above is a good guideline -- you don't have to use the full width and I don't want to start counting characters, but at the moment this block uses less than half 🙂 maybe 2/3-3/4 is ok.

Reply:
yeah this isn't new code, i can change it though